src/xviewer-list-store.c

```diff
@@ -627,9 +644,55 @@ xviewer_list_store_add_files (XviewerListStore *store, GList *file_list)
 			g_object_unref (file);
 		} else if (file_type == G_FILE_TYPE_REGULAR && !singleton_list) {
 			xviewer_list_store_append_image_from_file (store, file);
+
+			file = g_file_get_parent (file);
+			file_info = g_file_query_info (file,
+						       G_FILE_ATTRIBUTE_STANDARD_TYPE,
+						       0, NULL, NULL);
+
+			/* If we can't get a file_info,
+			   file_type will stay as G_FILE_TYPE_REGULAR */
+			if (file_info != NULL) {
+				file_type = g_file_info_get_file_type (file_info);
+				g_object_unref (file_info);
+			}
+                                                        /* attempt to set callbacks for the directories
+                                                           that contain the selected files - otherwise
+                                                           changes by external programs to an image that
+                                                           is currently being displayed will not cause
+                                                           the display to be updated */
+			if (file_type == G_FILE_TYPE_DIRECTORY) {
+                gboolean name_not_found = TRUE;
+
+                gchar * directory_name = g_file_get_basename(file);
+
+                dir_iter = directory_list;
+
+                while ((dir_iter != NULL) && name_not_found) {
+                    if (strcmp(dir_iter->data, directory_name) == 0)
+                        name_not_found = FALSE;
+                    dir_iter = dir_iter->next;
+                }
+
+                if (name_not_found) {
+                    directory_list = g_list_prepend(directory_list, g_strdup(directory_name));
+                    xviewer_list_store_set_directory_callbacks (store, file, file_type);
```

Comment:
`file` (the parent) can be `g_object_unref`'d here since it's not used anymore within this scope

Reply:
Unref of file added immediately after the g_free (directory_name); (just below the code shown in the above fragment).